{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import gc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch.nn as nn\n",
    "import torch "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Model Definition for scEGA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class graphattention_layer(nn.Module):\n",
    "    def __init__(self,input_size,numnodes,output_size,adjM):\n",
    "        #WkH_{i-1} is out dimension : CurrentNodeShape x N\n",
    "        self.inpshape = input_size\n",
    "        self.opshape = output_size\n",
    "        self.A = adjM\n",
    "        super.__init__(graphattention_layer)\n",
    "        self.vkt = nn.Linear(in_features=numnodes,output_size = 1)\n",
    "        self.vkr = nn.Linear(in_features=numnodes,output_size = 1)\n",
    "        self.W =  nn.Linear(in_features=input_size,out_features=output_size) \n",
    "    def forward(self, H_k):\n",
    "        #H_k represents the previous layer's graph representation\n",
    "        M_s = self.A * self.vkt(nn.ReLU(self.W(H_k))).T\n",
    "        M_r = (self.A * self.vkr(nn.ReLU(self.W(H_k))).T).T\n",
    "        Attention = nn.Softmax(nn.Sigmoid(M_s+M_r))\n",
    "        H_new = nn.ReLU(self.W(H_k))@Attention\n",
    "        return H_new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class encoder(nn.Module):\n",
    "    def __init__(self,hidden_size,adjM):\n",
    "        super.__init__(encoder)\n",
    "        self.layer1 = graphattention_layer(input_size=adjM.shape[0]\n",
    "                                           ,numnodes=adjM.shape(0),\n",
    "                                           output_size=(adjM.shape(0))/8,adjM=adjM)\n",
    "        self.layer2 = graphattention_layer(1,adjM)\n",
    "        self.layer3 = graphattention_layer(1,adjM)\n",
    "    def forward(self, H):\n",
    "        H1 = self.layer1(H)\n",
    "        H2 = self.layer2(H1)\n",
    "        H3 = self.layer3(H2)\n",
    "        return H3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class decoder(nn.Module):\n",
    "    def __init__(self,hidden_size,adjM,GeneGraph):\n",
    "        super.__init__(encoder)\n",
    "        self.gGraph = GeneGraph\n",
    "        self.layer1 = graphattention_layer(input_size=adjM.shape[0]\n",
    "                                           ,numnodes=adjM.shape(0),\n",
    "                                           output_size=(adjM.shape(0))/8,adjM=adjM)\n",
    "        self.layer2 = graphattention_layer(1,adjM)\n",
    "        self.layer3 = graphattention_layer(1,adjM)\n",
    "    def forward(self, H):\n",
    "        decodedPass = torch.stack(self.gGraph,H)\n",
    "        H1 = self.layer1(decodedPass)\n",
    "        H2 = self.layer2(H1)\n",
    "        H3 = self.layer3(H2)\n",
    "        return H3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class scdEGA(nn.Module):\n",
    "    def __init__(self,hidden_size,adjM,GeneGraph):\n",
    "        super.__init__(scdEGA)\n",
    "        self.encoder = encoder(hidden_size,adjM)\n",
    "        self.decoder = decoder(hidden_size,adjM,GeneGraph)\n",
    "    def forward(self, H):\n",
    "        H1 = self.encoder(H)\n",
    "        H2 = self.decoder(H1)\n",
    "        return H2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([[-0.1999],\n",
      "        [-0.3973],\n",
      "        [-0.8067],\n",
      "        [-0.7447],\n",
      "        [-0.5682],\n",
      "        [-0.4899],\n",
      "        [-0.5840],\n",
      "        [-0.6114],\n",
      "        [-0.6487],\n",
      "        [-0.4664]], grad_fn=<AddmmBackward0>)\n"
     ]
    }
   ],
   "source": [
    "check = nn.Linear(in_features=5,out_features=1)\n",
    "mat = torch.rand(size=(10,5)) # 5 nodes , each of a feature size of 10\n",
    "print(check(mat))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1:Data preprocessing\n",
    "\n",
    "This will involve the following steps:\n",
    "1. Obtaining a knn cell-cell graph\n",
    "2. Obtaining a gene-gene graph based on protein-protein interactomes from https://string-db.org/cgi/network?taskId=bSmGYnEnGRS8&sessionId=bG2dm5uLdSGY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1 : Obtaining Cell matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_text(\"./GSE57249_fpkm.txt\")\n",
    "adata = adata.T\n",
    "sc.pp.pca(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 56 × 25737\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb'\n",
       "    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    uns: 'pca'\n",
       "    obsm: 'X_pca'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 56 × 25737\n",
       "    obs: 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_20_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'log1p_total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb'\n",
       "    var: 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
       "    uns: 'pca'\n",
       "    obsm: 'X_pca'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mitochondrial genes\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "# ribosomal genes\n",
    "adata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\",\"RPL\"))\n",
    "# hemoglobin genes.\n",
    "adata.var[\"hb\"] = adata.var_names.str.contains((\"^HB[^(P)]\"))\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    adata, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, percent_top=[20], log1p=True\n",
    ")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "p1 = sns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n",
    "# sc.pl.violin(adata, 'total_counts')\n",
    "p2 = sc.pl.violin(adata, \"pct_counts_mt\")\n",
    "p3 = sc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Theres like no mtrna lol."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 = sns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n",
    "# sc.pl.violin(adata, 'total_counts')\n",
    "p2 = sc.pl.violin(adata, \"pct_counts_mt\")\n",
    "p3 = sc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Remove cells with a minimum number of expressed genes\n",
    "sc.pp.filter_cells(adata, min_genes=3)\n",
    "\n",
    "# Print the updated shape of the AnnData object\n",
    "print(\"Updated shape:\", adata.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "adj_list = adata.obsp[\"distances\"].indices.reshape(adata.shape[0],-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/saigum/anaconda3/envs/singleseq/lib/python3.12/site-packages/scanpy/preprocessing/_highly_variable_genes.py:305: RuntimeWarning: invalid value encountered in log\n",
      "  dispersion = np.log(dispersion)\n"
     ]
    }
   ],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=2000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/saigum/anaconda3/envs/singleseq/lib/python3.12/site-packages/scanpy/tools/_utils.py:41: UserWarning: You’re trying to run this on 56 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "sc.neighbors.neighbors(adata.T,5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our adjacency matrix in this case is simply our knn matrix, where we store the indices(\"is:cells\"), whose expressions are closest to the cell of that row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = torch.zeros(size=(adata.T.shape[1],adata.T.shape[1]))\n",
    "for i in range(adj_list.shape[0]):\n",
    "    for j in range(5):\n",
    "        A[i][adj_list[i][j]] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to obtain the gene-gene graph\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"highly_variable_genes.txt\", \"w\") as f:    \n",
    "     f.write(','.join(adata.var_names[adata.var[\"highly_variable\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_list = pd.DataFrame(csv.reader(open(\"string_interactions.tsv\"),delimiter=\"\\t\"))\n",
    "edge_list.columns = edge_list.iloc[0]\n",
    "edge_list = edge_list.drop(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>node1_string_id</th>\n",
       "      <th>node2_string_id</th>\n",
       "      <th>neighborhood_on_chromosome</th>\n",
       "      <th>gene_fusion</th>\n",
       "      <th>phylogenetic_cooccurrence</th>\n",
       "      <th>homology</th>\n",
       "      <th>coexpression</th>\n",
       "      <th>experimentally_determined_interaction</th>\n",
       "      <th>database_annotated</th>\n",
       "      <th>automated_textmining</th>\n",
       "      <th>combined_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610010F05Rik</td>\n",
       "      <td>Ahsa2</td>\n",
       "      <td>10090.ENSMUSP00000044265</td>\n",
       "      <td>10090.ENSMUSP00000020529</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.091</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1300017J02Rik</td>\n",
       "      <td>Trf</td>\n",
       "      <td>10090.ENSMUSP00000035163</td>\n",
       "      <td>10090.ENSMUSP00000035158</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.061</td>\n",
       "      <td>0.959</td>\n",
       "      <td>0.104</td>\n",
       "      <td>0</td>\n",
       "      <td>0.900</td>\n",
       "      <td>0</td>\n",
       "      <td>0.908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1300017J02Rik</td>\n",
       "      <td>Ica1</td>\n",
       "      <td>10090.ENSMUSP00000035163</td>\n",
       "      <td>10090.ENSMUSP00000040062</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.473</td>\n",
       "      <td>0.473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1600014C10Rik</td>\n",
       "      <td>Coasy</td>\n",
       "      <td>10090.ENSMUSP00000130271</td>\n",
       "      <td>10090.ENSMUSP00000102929</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.050</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1600014C10Rik</td>\n",
       "      <td>BC048679</td>\n",
       "      <td>10090.ENSMUSP00000130271</td>\n",
       "      <td>10090.ENSMUSP00000120616</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.604</td>\n",
       "      <td>0.604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3458</th>\n",
       "      <td>Zfp933</td>\n",
       "      <td>Rpp25</td>\n",
       "      <td>10090.ENSMUSP00000101343</td>\n",
       "      <td>10090.ENSMUSP00000079358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.073</td>\n",
       "      <td>0</td>\n",
       "      <td>0.388</td>\n",
       "      <td>0.408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3459</th>\n",
       "      <td>Zglp1</td>\n",
       "      <td>Figla</td>\n",
       "      <td>10090.ENSMUSP00000111157</td>\n",
       "      <td>10090.ENSMUSP00000032070</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.089</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.402</td>\n",
       "      <td>0.432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3460</th>\n",
       "      <td>Znrd1as</td>\n",
       "      <td>Rpp21</td>\n",
       "      <td>10090.ENSMUSP00000048695</td>\n",
       "      <td>10090.ENSMUSP00000025319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.079</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3461</th>\n",
       "      <td>Zswim2</td>\n",
       "      <td>Cct6b</td>\n",
       "      <td>10090.ENSMUSP00000044913</td>\n",
       "      <td>10090.ENSMUSP00000021040</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.511</td>\n",
       "      <td>0.517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3462</th>\n",
       "      <td>Zswim2</td>\n",
       "      <td>Saxo2</td>\n",
       "      <td>10090.ENSMUSP00000044913</td>\n",
       "      <td>10090.ENSMUSP00000057993</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.497</td>\n",
       "      <td>0.527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3462 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0            #node1     node2           node1_string_id  \\\n",
       "1     0610010F05Rik     Ahsa2  10090.ENSMUSP00000044265   \n",
       "2     1300017J02Rik       Trf  10090.ENSMUSP00000035163   \n",
       "3     1300017J02Rik      Ica1  10090.ENSMUSP00000035163   \n",
       "4     1600014C10Rik     Coasy  10090.ENSMUSP00000130271   \n",
       "5     1600014C10Rik  BC048679  10090.ENSMUSP00000130271   \n",
       "...             ...       ...                       ...   \n",
       "3458         Zfp933     Rpp25  10090.ENSMUSP00000101343   \n",
       "3459          Zglp1     Figla  10090.ENSMUSP00000111157   \n",
       "3460        Znrd1as     Rpp21  10090.ENSMUSP00000048695   \n",
       "3461         Zswim2     Cct6b  10090.ENSMUSP00000044913   \n",
       "3462         Zswim2     Saxo2  10090.ENSMUSP00000044913   \n",
       "\n",
       "0              node2_string_id neighborhood_on_chromosome gene_fusion  \\\n",
       "1     10090.ENSMUSP00000020529                          0           0   \n",
       "2     10090.ENSMUSP00000035158                          0           0   \n",
       "3     10090.ENSMUSP00000040062                          0           0   \n",
       "4     10090.ENSMUSP00000102929                          0           0   \n",
       "5     10090.ENSMUSP00000120616                          0           0   \n",
       "...                        ...                        ...         ...   \n",
       "3458  10090.ENSMUSP00000079358                          0           0   \n",
       "3459  10090.ENSMUSP00000032070                          0           0   \n",
       "3460  10090.ENSMUSP00000025319                          0           0   \n",
       "3461  10090.ENSMUSP00000021040                          0           0   \n",
       "3462  10090.ENSMUSP00000057993                          0           0   \n",
       "\n",
       "0    phylogenetic_cooccurrence homology coexpression  \\\n",
       "1                            0        0        0.091   \n",
       "2                        0.061    0.959        0.104   \n",
       "3                            0        0            0   \n",
       "4                            0        0        0.050   \n",
       "5                            0        0            0   \n",
       "...                        ...      ...          ...   \n",
       "3458                         0        0            0   \n",
       "3459                         0        0        0.089   \n",
       "3460                         0        0        0.079   \n",
       "3461                         0        0        0.052   \n",
       "3462                         0        0        0.099   \n",
       "\n",
       "0    experimentally_determined_interaction database_annotated  \\\n",
       "1                                        0                  0   \n",
       "2                                        0              0.900   \n",
       "3                                        0                  0   \n",
       "4                                        0                  0   \n",
       "5                                        0                  0   \n",
       "...                                    ...                ...   \n",
       "3458                                 0.073                  0   \n",
       "3459                                     0                  0   \n",
       "3460                                     0                  0   \n",
       "3461                                     0                  0   \n",
       "3462                                     0                  0   \n",
       "\n",
       "0    automated_textmining combined_score  \n",
       "1                   0.392          0.423  \n",
       "2                       0          0.908  \n",
       "3                   0.473          0.473  \n",
       "4                   0.538          0.542  \n",
       "5                   0.604          0.604  \n",
       "...                   ...            ...  \n",
       "3458                0.388          0.408  \n",
       "3459                0.402          0.432  \n",
       "3460                0.538          0.556  \n",
       "3461                0.511          0.517  \n",
       "3462                0.497          0.527  \n",
       "\n",
       "[3462 rows x 13 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_graph = nx.Graph()\n",
    "gene_graph.add_edges_from(edge_list[[\"node1_string_id\",\"node2_string_id\"]].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from node2vec import Node2Vec\n",
    "node2vec = Node2Vec(gene_graph, dimensions=64, walk_length=30, num_walks=200, workers=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for i in range(edge_list.shape[0]):\n",
    "    gene_graph.add_edge(edge_list., edge_list.iloc[i,1])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "singleseq",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
